File: stage-common/src/main/java/io/github/solusmods/eternalcore/stage/api/IReachedStages.java
package io.github.solusmods.eternalcore.stage.api;

import lombok.NonNull;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;

public interface IReachedStages {

    Collection<StageInstance> getReachedStages();

    default boolean addStage(@NotNull ResourceLocation stageId, boolean teleportToSpawn) {
        return addStage(stageId, teleportToSpawn, null);
    }

    default boolean addStage(@NotNull ResourceLocation stageId, boolean teleportToSpawn, @Nullable MutableComponent component) {
        Stage stage = StageAPI.getStageRegistry().get(stageId);
        if (stage == null) return false;
        return addStage(stage.createDefaultInstance(), false, teleportToSpawn, component);
    }

    default boolean addStage(@NonNull Stage stage, boolean teleportToSpawn) {
        return addStage(stage, teleportToSpawn, null);
    }

    default boolean addStage(@NonNull Stage stage, boolean teleportToSpawn, @Nullable MutableComponent component) {
        return addStage(stage.createDefaultInstance(), false, teleportToSpawn, component);
    }

    default boolean addStage(StageInstance instance, boolean breakthrough, boolean teleportToSpawn) {
        return addStage(instance, breakthrough, teleportToSpawn, null);
    }

    boolean addStage(StageInstance instance, boolean breakthrough, boolean teleportToSpawn, @Nullable MutableComponent component);

    void markDirty();

    void sync();
}

File: stage-common/src/main/kotlin/io/github/solusmods/eternalcore/stage/api/IReachedStages.kt
package io.github.solusmods.eternalcore.stage.api

import net.minecraft.network.chat.MutableComponent
import net.minecraft.resources.ResourceLocation

interface IReachedStages {
    val reachedStages: MutableMap<ResourceLocation?, StageInstance>


    fun addStage(stageId: ResourceLocation, notify: Boolean?, component: MutableComponent? = null): Boolean {
        val stage = StageAPI.stageRegistry!!.get(stageId)
        if (stage == null) return false
        return addStage(stage.createDefaultInstance(), false, notify!!, component)
    }


    fun addStage(stage: Stage, notify: Boolean?, component: MutableComponent? = null): Boolean {
        return addStage(stage.createDefaultInstance(), false, notify!!, component)
    }

    fun addStage(instance: StageInstance, breakthrough: Boolean?, notify: Boolean): Boolean {
        return addStage(instance, breakthrough, notify, null)
    }

    fun addStage(
        instance: StageInstance,
        breakthrough: Boolean?,
        notify: Boolean?,
        component: MutableComponent?
    ): Boolean

    fun markDirty()

    fun sync()
}
